### PR TITLE
Add _pyodide_core C extension module

### PR DESCRIPTION
**Pull request by hoodmane**
This is meant to address two issues: 
1. hard to tell which things are defined in core #1079,
2. the bootstrap order is confusing -- `pyodide_py` needs a most of the stuff from `core` in order to work, but `core` imports `pyodide_py` fairly early on in the initialization. Currently it works fine because `pyodide_py` only needs `core` to run it's methods not for initialization, but it's easy to make innocuous-seeming changes that mess it up.

Instead of directly adding stuff to `pyodide_py`, we make a C module called `_core` and add things to this instead. I also added `_core.py` which imports names from `_core` if we are running in the browser and spits out empty shims otherwise. If needed for testing reasons, these shims can be modified. In `_core.py` we can specify which C source file added each object imported from `_core`.

### CODE REVIEW

File: src/tests/test_pyproxy.py [hunk 1]
```diff
@@ -62,6 +62,49 @@ def get_value(self, value):
     assert selenium.run_js("return pyodide.pyimport('f').toString()").startswith("<Foo")
 
 
+def test_pyproxy2(selenium):
```
Comment (rth) [inline src/tests/test_pyproxy.py:65]: The test name needs to be meaningful, here and below

File: src/pyodide-py/pyodide/_core.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# type: ignore
+import platform
+
+if platform.system() == "Emscripten":
+    from _core import JsProxy, JsBoundMethod, JsException
```
Comment (rth) [inline src/pyodide-py/pyodide/_core.py:5]: It should have pyodide in its import name. Otherwise it's not clear at all that `_core` has anything to do with pyodide. Maybe `_pyodide_core` ?

File: src/pyodide-py/pyodide/_core.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+else:
+    # Can add shims here if we are so inclined.
+    class JsException(Exception):
```
Comment (rth) [inline src/pyodide-py/pyodide/_core.py:8]: If it only contains exception maybe put it in `pyodide/_exceptions.py` ?

File: src/core/jsproxy.c [hunk 4]
```diff
@@ -696,27 +696,39 @@ JsException_AsJs(PyObject* err)
   return hiwire_incref(js_error->js);
 }
 
+// Copied from Python 3.9
```
Comment (rth) [inline src/core/jsproxy.c:699]:
```suggestion
// Copied from Python 3.9
// TODO: remove once we update to Python 3.9
```
If I understand correctly?

File: src/tests/test_pyproxy.py [hunk 5]
```diff
@@ -62,6 +62,49 @@ def get_value(self, value):
...
+        """
+        pyodide.runPython(`
+            class Foo:
+                pass
+            f = Foo()
+        `);
+        return pyodide.pyimport('f').toString();
+        """
+    )
+    print(result)
```
Comment (rth) [inline src/tests/test_pyproxy.py:76]:
```suggestion

```

File: src/tests/test_pyproxy.py [hunk 6]
```diff
@@ -62,6 +62,49 @@ def get_value(self, value):
...
+    assert result.startswith("<Foo")
+
+
+def test_pyproxy3(selenium):
+    selenium.run_js(
+        """
+        window.jspython = pyodide.pyimport("open");
+        pyodide.runPython(`
+            from js import jspython
+            jspython is open
```
Comment (rth) [inline src/tests/test_pyproxy.py:86]:
```suggestion
            assert jspython is open
```
?
Is this related to this PR though? Maybe it would be better to add theses tests in a separate PR?

File: src/tests/test_pyproxy.py [hunk 7]
```diff
@@ -62,6 +62,49 @@ def get_value(self, value):
...
+        """
+        window.jspython = pyodide.pyimport("open");
+        console.log("jspython:", window.jspython)
+        """
+    )
+    assert selenium.run(
+        """
+        from js import jspython
+        print(jspython, open)
+        jspython is open
```
Comment (rth) [inline src/tests/test_pyproxy.py:103]:
```suggestion
        assert jspython is open
```

File: src/core/jsproxy.c (hunk 4 above)
Comment (hoodmane) [reply src/core/jsproxy.c:699]: Absolutely, it'll lead to a linker error otherwise.

File: src/pyodide-py/pyodide/_core.py (hunk 3 above)
Comment (hoodmane) [reply src/pyodide-py/pyodide/_core.py:8]: `JsProxy` and `JsBoundMethod` aren't exceptions? Also I'm planning to automatically add `console` into this as well. `console.error` produces incredibly more helpful messages than `print`, even when just applied to python objects.

File: src/tests/test_pyproxy.py (hunk 1 above)
Comment (hoodmane) [reply src/tests/test_pyproxy.py:65]: Yes of course =) I swear I meant to rename them. When I first name them, I'm thinking more about whether `pytest -k whatever` will get them and the test I already had.

This test is actually just a redundant minimization of an existing test, so either this should be removed or the other test should be broken up into smaller pieces.

File: src/tests/test_pyproxy.py (hunk 6 above)
Comment (hoodmane) [reply src/tests/test_pyproxy.py:86]: Again this is a minimization of an existing test that was breaking and should be removed in any case (it did lead to PR #1098).

File: src/pyodide-py/pyodide/_core.py (hunk 2 above)
Comment (hoodmane) [reply src/pyodide-py/pyodide/_core.py:5]: Sounds good.

**Review by rth — APPROVED** [review]
Thanks !
